File: static/remote-machines.js
/**
 * Copyright (c) 2025 BillChen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * 远程机器管理模块
 * 提供远程 SSH 机器的添加、编辑、删除、连接功能
 */
const RemoteMachines = {
  machines: [],
  editingMachineId: null,

  /**
   * 初始化模块
   */
  init() {
    this.bindEvents();
  },

  /**
   * 绑定事件
   */
  bindEvents() {
    // 添加远程机器按钮
    const addBtn = document.getElementById('add-remote-machine-btn');
    if (addBtn) {
      addBtn.addEventListener('click', () => this.openAddModal());
    }

    // 关闭模态框
    const closeBtn = document.getElementById('remote-machine-modal-close');
    if (closeBtn) {
      closeBtn.addEventListener('click', () => this.closeModal());
    }

    // 测试连接按钮
    const testBtn = document.getElementById('test-connection-btn');
    if (testBtn) {
      testBtn.addEventListener('click', () => this.testConnection());
    }

    // 保存按钮
    const saveBtn = document.getElementById('save-remote-machine-btn');
    if (saveBtn) {
      saveBtn.addEventListener('click', () => this.saveMachine());
    }

    // 点击模态框背景关闭
    const modal = document.getElementById('remote-machine-modal');
    if (modal) {
      modal.addEventListener('click', (e) => {
        if (e.target === modal) {
          this.closeModal();
        }
      });
    }
  },

  /**
   * 加载远程机器列表
   */
  async loadMachines() {
    const container = document.getElementById('remote-machines-list');
    if (!container) return;

    container.innerHTML = `<div class="loading">${this.t('sessions.loading')}</div>`;

    try {
      const response = await fetch('/api/remote-machines', {
        headers: {
          'Authorization': `Bearer ${this.token}`
        }
      });

      if (response.status === 401) {
        this.handleUnauthorized();
        return;
      }

      if (!response.ok) throw new Error('Failed to load remote machines');

      this.machines = await response.json();

      if (this.machines.length === 0) {
        container.innerHTML = `
          <div class="no-sessions">
            <div>${this.t('remote.noMachines')}</div>
          </div>
        `;
        return;
      }

      this.renderMachineList(container);
    } catch (error) {
      console.error('Load remote machines error:', error);
      container.innerHTML = `<div class="no-sessions">${this.t('sessions.loadFailed')}</div>`;
    }
  },

  /**
   * 渲染机器列表
   */
  renderMachineList(container) {
    container.innerHTML = '';

    this.machines.forEach(machine => {
      const item = document.createElement('div');
      item.className = 'remote-machine-item';
      item.innerHTML = `
        <div class="machine-info" data-id="${machine.id}">
          <div class="machine-icon">>_</div>
          <div class="machine-details">
            <div class="machine-name">${this.escapeHtml(machine.name)}</div>
            <div class="machine-host">${this.escapeHtml(machine.username)}@${this.escapeHtml(machine.host)}:${machine.port}</div>
          </div>
        </div>
        <div class="machine-actions">
          <button class="action-btn edit-btn" onclick="RemoteMachines.openEditModal(${machine.id})" title="${this.t('remote.edit')}">✎</button>
          <button class="action-btn delete-btn" onclick="RemoteMachines.deleteMachine(${machine.id})" title="${this.t('remote.delete')}">✕</button>
        </div>
      `;

      // 点击卡片连接
      const infoDiv = item.querySelector('.machine-info');
      infoDiv.addEventListener('click', () => this.connectToMachine(machine));

      container.appendChild(item);
    });

    // 强制 reflow，解决离屏页面不渲染的问题
    container.offsetHeight;
  },

  /**
   * 打开添加模态框
   */
  openAddModal() {
    this.editingMachineId = null;
    this.clearForm();
    document.getElementById('remote-machine-modal-title').textContent = this.t('remote.addMachine');
    document.getElementById('remote-machine-modal').classList.add('active');
  },

  /**
   * 打开编辑模态框
   */
  async openEditModal(machineId) {
    this.editingMachineId = machineId;
    const machine = this.machines.find(m => m.id === machineId);
    if (!machine) return;

    document.getElementById('machine-name').value = machine.name;
    document.getElementById('machine-host').value = machine.host;
    document.getElementById('machine-port').value = machine.port;
    document.getElementById('machine-username').value = machine.username;
    document.getElementById('machine-password').value = ''; // 不显示旧密码
    document.getElementById('machine-password').placeholder = this.t('remote.passwordPlaceholder');

    document.getElementById('remote-machine-modal-title').textContent = this.t('remote.editMachine');
    document.getElementById('remote-machine-modal').classList.add('active');
  },

  /**
   * 关闭模态框
   */
  closeModal() {
    document.getElementById('remote-machine-modal').classList.remove('active');
    this.clearForm();
    this.editingMachineId = null;
  },

  /**
   * 清空表单
   */
  clearForm() {
    document.getElementById('machine-name').value = '';
    document.getElementById('machine-host').value = '';
    document.getElementById('machine-port').value = '22';
    document.getElementById('machine-username').value = '';
    document.getElementById('machine-password').value = '';
    document.getElementById('machine-password').placeholder = '';
    document.getElementById('test-connection-result').textContent = '';
    document.getElementById('test-connection-result').className = 'test-result';
  },

  /**
   * 测试连接
   */
  async testConnection() {
    const resultDiv = document.getElementById('test-connection-result');
    resultDiv.textContent = this.t('remote.testing');
    resultDiv.className = 'test-result testing';

    const data = this.getFormData();
    if (!data) {
      resultDiv.textContent = this.t('remote.fillRequired');
      resultDiv.className = 'test-result error';
      return;
    }

    try {
      const response = await fetch('/api/remote-machines/test', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.token}`
        },
        body: JSON.stringify(data)
      });

      const result = await response.json();

      if (result.success) {
        resultDiv.textContent = this.t('remote.testSuccess');
        resultDiv.className = 'test-result success';
      } else {
        resultDiv.textContent = `${this.t('remote.testFailed')}: ${result.message}`;
        resultDiv.className = 'test-result error';
      }
    } catch (error) {
      console.error('Test connection error:', error);
      resultDiv.textContent = this.t('remote.testFailed');
      resultDiv.className = 'test-result error';
    }
  },

  /**
   * 保存机器
   */
  async saveMachine() {
    const data = this.getFormData();
    if (!data) {
      window.app.showAlert(this.t('remote.fillRequired'), { type: 'warning' });
      return;
    }

    // 编辑模式下，如果密码为空则不更新密码
    if (this.editingMachineId && !data.password) {
      delete data.password;
    }

    try {
      const url = this.editingMachineId
        ? `/api/remote-machines/${this.editingMachineId}`
        : '/api/remote-machines';
      const method = this.editingMachineId ? 'PUT' : 'POST';

      const response = await fetch(url, {
        method,
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.token}`
        },
        body: JSON.stringify(data)
      });

      if (!response.ok) throw new Error('Save failed');

      this.closeModal();
      await this.loadMachines();
    } catch (error) {
      console.error('Save machine error:', error);
      window.app.showAlert(this.t('remote.saveFailed'), { type: 'error' });
    }
  },

  /**
   * 删除机器
   */
  async deleteMachine(machineId) {
    const machine = this.machines.find(m => m.id === machineId);
    if (!machine) return;

    const confirmed = await window.app.showConfirm(`${this.t('remote.confirmDelete')} "${machine.name}"?`, {
      type: 'danger',
      title: this.t('dialog.delete', 'Delete'),
      confirmText: this.t('common.delete', 'Delete')
    });
    if (!confirmed) return;

    try {
      const response = await fetch(`/api/remote-machines/${machineId}`, {
        method: 'DELETE',
        headers: {
          'Authorization': `Bearer ${this.token}`
        }
      });

      if (!response.ok) throw new Error('Delete failed');

      await this.loadMachines();
    } catch (error) {
      console.error('Delete machine error:', error);
      window.app.showAlert(this.t('remote.deleteFailed'), { type: 'error' });
    }
  },

  /**
   * 连接到远程机器
   */
  connectToMachine(machine) {
    // 调用独立的 SSH 终端模块
    if (window.SSHTerminal && typeof window.SSHTerminal.connect === 'function') {
      window.SSHTerminal.connect(machine);
    } else {
      console.error('SSHTerminal module not found');
    }
  },

  /**
   * 获取表单数据
   */
  getFormData() {
    const name = document.getElementById('machine-name').value.trim();
    const host = document.getElementById('machine-host').value.trim();
    const port = parseInt(document.getElementById('machine-port').value) || 22;
    const username = document.getElementById('machine-username').value.trim();
    const password = document.getElementById('machine-password').value;

    // 验证必填字段
    if (!name || !host || !username) {
      return null;
    }

    // 新增时密码必填
    if (!this.editingMachineId && !password) {
      return null;
    }

    return { name, host, port, username, password };
  },

  /**
   * HTML 转义
   */
  escapeHtml(text) {
    const div = document.createElement('div');
    div.textContent = text;
    return div.innerHTML;
  },

  /**
   * 国际化翻译
   */
  t(key) {
    if (window.i18n && typeof i18n.t === 'function') {
      return i18n.t(key);
    }
    // 默认英文
    const defaults = {
      'sessions.loading': 'Loading...',
      'sessions.loadFailed': 'Load failed',
      'remote.noMachines': 'No remote machines configured',
      'remote.addMachine': 'Add Remote Machine',
      'remote.editMachine': 'Edit Remote Machine',
      'remote.edit': 'Edit',
      'remote.delete': 'Delete',
      'remote.testing': 'Testing connection...',
      'remote.testSuccess': 'Connection successful!',
      'remote.testFailed': 'Connection failed',
      'remote.fillRequired': 'Please fill in all required fields',
      'remote.saveFailed': 'Save failed',
      'remote.deleteFailed': 'Delete failed',
      'remote.confirmDelete': 'Are you sure you want to delete',
      'remote.passwordPlaceholder': 'Leave empty to keep current password'
    };
    return defaults[key] || key;
  },

  /**
   * 处理未授权
   */
  handleUnauthorized() {
    if (window.app && typeof window.app.handleUnauthorized === 'function') {
      window.app.handleUnauthorized();
    }
  },

  /**
   * 获取 token
   */
  get token() {
    if (window.app && window.app.token) {
      return window.app.token;
    }
    return localStorage.getItem('token') || '';
  }
};

// 导出到全局
window.RemoteMachines = RemoteMachines;

File: app/api/remote_machines.py
# Copyright (c) 2025 BillChen
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Remote Machines API

管理远程 SSH 机器的 CRUD 接口。
"""
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import List, Optional

from app.api.auth import verify_token
from app.services.database import db
from app.core.logging import logger

router = APIRouter(prefix="/api/remote-machines", tags=["remote-machines"])


# ==================== Request/Response Models ====================

class RemoteMachineCreate(BaseModel):
    name: str
    host: str
    port: int = 22
    username: str
    password: str
    auth_type: str = "password"


class RemoteMachineUpdate(BaseModel):
    name: Optional[str] = None
    host: Optional[str] = None
    port: Optional[int] = None
    username: Optional[str] = None
    password: Optional[str] = None
    auth_type: Optional[str] = None


class RemoteMachineResponse(BaseModel):
    id: int
    name: str
    host: str
    port: int
    username: str
    auth_type: str
    created_at: str
    updated_at: str


class TestConnectionRequest(BaseModel):
    host: str
    port: int = 22
    username: str
    password: str


class TestConnectionResponse(BaseModel):
    success: bool
    message: str


# ==================== API Endpoints ====================

@router.get("", response_model=List[RemoteMachineResponse])
async def list_remote_machines(_: str = Depends(verify_token)):
    """获取所有远程机器列表"""
    try:
        machines = db.get_remote_machines()
        return [
            RemoteMachineResponse(
                id=m["id"],
                name=m["name"],
                host=m["host"],
                port=m["port"],
                username=m["username"],
                auth_type=m["auth_type"],
                created_at=m["created_at"],
                updated_at=m["updated_at"]
            )
            for m in machines
        ]
    except Exception as e:
        logger.error(f"List remote machines error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{machine_id}", response_model=RemoteMachineResponse)
async def get_remote_machine(machine_id: int, _: str = Depends(verify_token)):
    """获取单个远程机器信息"""
    try:
        machine = db.get_remote_machine(machine_id)
        if not machine:
            raise HTTPException(status_code=404, detail="Remote machine not found")
        return RemoteMachineResponse(
            id=machine["id"],
            name=machine["name"],
            host=machine["host"],
            port=machine["port"],
            username=machine["username"],
            auth_type=machine["auth_type"],
            created_at=machine["created_at"],
            updated_at=machine["updated_at"]
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Get remote machine error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("", response_model=RemoteMachineResponse)
async def create_remote_machine(
    data: RemoteMachineCreate,
    _: str = Depends(verify_token)
):
    """添加远程机器"""
    try:
        machine_id = db.add_remote_machine(
            name=data.name,
            host=data.host,
            port=data.port,
            username=data.username,
            password=data.password,
            auth_type=data.auth_type
        )
        machine = db.get_remote_machine(machine_id)
        return RemoteMachineResponse(
            id=machine["id"],
            name=machine["name"],
            host=machine["host"],
            port=machine["port"],
            username=machine["username"],
            auth_type=machine["auth_type"],
            created_at=machine["created_at"],
            updated_at=machine["updated_at"]
        )
    except Exception as e:
        logger.error(f"Create remote machine error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/{machine_id}", response_model=RemoteMachineResponse)
async def update_remote_machine(
    machine_id: int,
    data: RemoteMachineUpdate,
    _: str = Depends(verify_token)
):
    """更新远程机器信息"""
    try:
        # 检查是否存在
        existing = db.get_remote_machine(machine_id)
        if not existing:
            raise HTTPException(status_code=404, detail="Remote machine not found")

        # 更新
        success = db.update_remote_machine(
            machine_id=machine_id,
            name=data.name,
            host=data.host,
            port=data.port,
            username=data.username,
            password=data.password,
            auth_type=data.auth_type
        )
        if not success:
            raise HTTPException(status_code=500, detail="Update failed")

        machine = db.get_remote_machine(machine_id)
        return RemoteMachineResponse(
            id=machine["id"],
            name=machine["name"],
            host=machine["host"],
            port=machine["port"],
            username=machine["username"],
            auth_type=machine["auth_type"],
            created_at=machine["created_at"],
            updated_at=machine["updated_at"]
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Update remote machine error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{machine_id}")
async def delete_remote_machine(machine_id: int, _: str = Depends(verify_token)):
    """删除远程机器"""
    try:
        success = db.delete_remote_machine(machine_id)
        if not success:
            raise HTTPException(status_code=404, detail="Remote machine not found")
        return {"success": True, "message": "Remote machine deleted"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Delete remote machine error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/{machine_id}/test", response_model=TestConnectionResponse)
async def test_connection_by_id(machine_id: int, _: str = Depends(verify_token)):
    """测试已保存的远程机器连接"""
    try:
        machine = db.get_remote_machine(machine_id)
        if not machine:
            raise HTTPException(status_code=404, detail="Remote machine not found")

        password = db.get_remote_machine_password(machine_id)

        # 延迟导入，避免循环依赖
        from app.services.ssh_manager import ssh_manager
        success, message = await ssh_manager.test_connection(
            host=machine["host"],
            port=machine["port"],
            username=machine["username"],
            password=password
        )
        return TestConnectionResponse(success=success, message=message)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Test connection error: {e}")
        return TestConnectionResponse(success=False, message=str(e))


@router.post("/test", response_model=TestConnectionResponse)
async def test_connection(data: TestConnectionRequest, _: str = Depends(verify_token)):
    """测试新的远程机器连接（不保存）"""
    try:
        from app.services.ssh_manager import ssh_manager
        success, message = await ssh_manager.test_connection(
            host=data.host,
            port=data.port,
            username=data.username,
            password=data.password
        )
        return TestConnectionResponse(success=success, message=message)
    except Exception as e:
        logger.error(f"Test connection error: {e}")
        return TestConnectionResponse(success=False, message=str(e))


@router.post("/{machine_id}/pin")
async def pin_ssh_session(machine_id: int, _: str = Depends(verify_token)):
    """将 SSH 会话固定到会话列表"""
    try:
        # 检查机器是否存在
        machine = db.get_remote_machine(machine_id)
        if not machine:
            raise HTTPException(status_code=404, detail="Remote machine not found")

        # 检查是否已固定
        if db.is_ssh_session_pinned(machine_id):
            return {"success": True, "message": "Already pinned", "already_pinned": True}

        # 添加固定
        result = db.add_ssh_pinned_session(
            machine_id=machine_id,
            machine_name=machine["name"]
        )

        if result is None:
            return {"success": True, "message": "Already pinned", "already_pinned": True}

        return {"success": True, "id": result, "message": "Pinned successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Pin SSH session error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{machine_id}/pin")
async def unpin_ssh_session(machine_id: int, _: str = Depends(verify_token)):
    """取消固定 SSH 会话"""
    try:
        session_id = f"ssh_{machine_id}"
        success = db.remove_pinned_session(session_id)
        if not success:
            raise HTTPException(status_code=404, detail="Session not pinned")
        return {"success": True, "message": "Unpinned successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Unpin SSH session error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{machine_id}/pin")
async def check_ssh_pinned(machine_id: int, _: str = Depends(verify_token)):
    """检查 SSH 会话是否已固定"""
    try:
        is_pinned = db.is_ssh_session_pinned(machine_id)
        return {"pinned": is_pinned}
    except Exception as e:
        logger.error(f"Check SSH pinned error: {e}")
        raise HTTPException(status_code=500, detail=str(e))
